### ISSUE TITLE
Direct access to sendgrid Mail object

### ISSUE THREAD

**Issue by davegaeddert** [elbuo8/sendgrid-django#45]
I'm wanting to change some additional settings on the sendgrid Mail object before it's sent (specifically click tracking - https://github.com/sendgrid/sendgrid-python/blob/master/sendgrid/helpers/mail/mail.py#L23).

I was wondering if there is or should be a more generic way to access the underlying object that comes out of [_build_sg_email](https://github.com/elbuo8/sendgrid-django/blob/master/sgbackend/mail.py#L75)? Suggestions for how to accomplish that (happy to help do it)?

Thanks for your work on this!

**Comment by andriisoldatenko** [comment]
@davegaeddert Thanks for your message. If I understand correctly, you want to add feature, to 
use `tracking_settings` mail object when you send email? I think it's possible.
@elbuo8 How do you think?
@thinkingserious

**Comment by elbuo8** [comment]
We would have to add a section to handle it on `_build_sg_email`. As it stands right now it won't copy it.

**Comment by davegaeddert** [comment]
Right, I guess what I'm wondering is if there's a more generic way to do it, so that you don't have to add lines like these for each property that just gets passed on to the sendgrid object:
- https://github.com/elbuo8/sendgrid-django/blob/master/sgbackend/mail.py#L103
- https://github.com/elbuo8/sendgrid-django/blob/master/sgbackend/mail.py#L109

The ones you have already seem handy, but it would be nice if there was a way to simply access the `Mail` object and set any advanced settings directly on it myself. Does that make sense? Then people could also mess with any of the other properties available here without you having to specifically add support for it: https://github.com/sendgrid/sendgrid-python/blob/master/sendgrid/helpers/mail/mail.py#L8

I'm not sure what the best way to do it would be though, since it looks like the email backend code doesn't get called until send?

**Comment by elbuo8** [comment]
Only thing that comes to mind is patching the `send_messages` which is not generic. 

We could make a whitelist and copy properties to the mail object too.

**Comment by davegaeddert** [comment]
Maybe a whitelist of properties that can all be copied is as good as anything. The other thing that came to mind for me, which might not be a good pattern, would be to pass a function that could get called. Haven't tried any of this but something like:
```python
mail = EmailMultiAlternatives(
  subject="Your Subject",
  body="This is a simple text email body.",
  from_email="Yamil Asusta <hello@yamilasusta.com>",
  to=["yamil@sendgrid.com"],
  headers={"Reply-To": "support@sendgrid.com"}
)
def sendgrid_mail_func(mail):
  mail.tracking_settings = 'anything you need'
  # etc.

mail.modify_sendgrid_mail_func = sendgrid_mail_func
```
Then in your backend
```python
if hasattr(email, 'modify_sendgrid_mail_func'):
  email.modify_sendgrid_mail_func(mail)
```

**Comment by andriisoldatenko** [comment]
>We could make a whitelist and copy properties to the mail object too.
+1
I think this approach more safety and guarantee that everything will work.

**Comment by davegaeddert** [comment]
@andriisoldatenko Sounds fine to me. Is that something that you guys are open to tackling relatively soon?

**Comment by estianross** [comment]
Is there any update on this being tackled? Specifically the handling of **mail.tracking_settings**?

**Comment by davegaeddert** [comment]
Personally I don't need this anymore (trying to close some old issues...). Feel free to reopen if somebody actually cares about this still. Thanks!
